app/helpers/i18n_helper.rb

```diff
@@ -24,9 +24,23 @@ def valid_locale(user)
     end
   end
 
-  private
-
   def available_locale?(locale)
     Rails.application.config.i18n.available_locales.include?(locale)
   end
+
+  private
+
+  def current_user_locale
+    spree_current_user.andand.locale
+  end
+
+  def valid_current_locale
+    if available_locale?(current_user_locale)
+      current_user_locale
+    elsif available_locale?(cookies[:locale])
+      cookies[:locale]
```

Comment:
if the cookies logic above is changed you dont need this because you have already saved cookies locale to the user.

Reply:
This method needs to return the most reasonable valid locale for the current page request, but the user may or may not be logged in. If they're not logged in, I think we we do need to look at the `cookies[:locale]` value, as a lower priority than the user's locale (if one is present), and higher priority than returning `I18n.default_locale`.

Reply:
ok, I think this makes sense.